File: src/loomcli/agents/anthropic.py
import json
import httpx
from typing import Generator, List, Dict, Any, Optional
from .base import AIProvider


class AnthropicProvider(AIProvider):
    def __init__(self, api_key: str):
        self.api_key = api_key
        self.base_url = "https://api.anthropic.com/v1/messages"
        self.api_version = "2023-06-01"

    def _convert_messages(self, messages: List[Dict]) -> tuple:
        system = None
        converted = []
        for msg in messages:
            role = msg["role"]
            if role == "system":
                system = msg["content"]
                continue
            content = msg.get("content") or ""
            if role == "tool":
                converted.append({
                    "role": "user",
                    "content": [{
                        "type": "tool_result",
                        "tool_use_id": msg.get("tool_call_id", ""),
                        "content": msg.get("content", "")
                    }]
                })
            elif role == "assistant":
                content_parts = []
                if content:
                    content_parts.append({"type": "text", "text": content})
                for tc in msg.get("tool_calls", []):
                    try:
                        args = json.loads(tc["function"]["arguments"]) if isinstance(tc["function"]["arguments"], str) else tc["function"]["arguments"]
                    except Exception:
                        args = tc["function"]["arguments"]
                    content_parts.append({
                        "type": "tool_use",
                        "id": tc["id"],
                        "name": tc["function"]["name"],
                        "input": args
                    })
                converted.append({"role": "assistant", "content": content_parts})
            else:
                converted.append({"role": role, "content": content})
        return system, converted

    def _convert_tools(self, tools: List[Dict]) -> List[Dict]:
        result = []
        for t in tools:
            fn = t.get("function", {})
            result.append({
                "name": fn.get("name", ""),
                "description": fn.get("description", ""),
                "input_schema": fn.get("parameters", {})
            })
        return result

    def ask(self, messages: List[Dict[str, str]], model: str, stream: bool = True, tools: Optional[List[Dict[str, Any]]] = None) -> Generator[Dict[str, Any], None, None]:
        system, converted = self._convert_messages(messages)

        payload = {
            "model": model,
            "messages": converted,
            "max_tokens": 4096,
            "stream": stream
        }
        if system:
            payload["system"] = system
        if tools:
            payload["tools"] = self._convert_tools(tools)

        headers = {
            "x-api-key": self.api_key,
            "anthropic-version": self.api_version,
            "Content-Type": "application/json"
        }

        with httpx.stream("POST", self.base_url, headers=headers, json=payload, timeout=60.0) as response:
            if response.status_code != 200:
                error_body = response.read().decode()
                yield {"type": "error", "content": f"Error from Anthropic ({response.status_code}): {error_body}"}
                return

            current_tool_block = None

            for line in response.iter_lines():
                if not line:
                    continue
                if line.startswith("event: "):
                    continue
                if not line.startswith("data: "):
                    continue

                data = line[6:]
                if data == "[DONE]":
                    break

                try:
                    event = json.loads(data)
                except json.JSONDecodeError:
                    continue

                event_type = event.get("type", "")

                if event_type == "content_block_start":
                    block = event.get("content_block", {})
                    if block.get("type") == "tool_use":
                        current_tool_block = {
                            "id": block.get("id", ""),
                            "type": "function",
                            "function": {
                                "name": block.get("name", ""),
                                "arguments": ""
                            }
                        }

                elif event_type == "content_block_delta":
                    delta = event.get("delta", {})
                    if delta.get("type") == "text_delta":
                        yield {"type": "text", "content": delta.get("text", "")}
                    elif delta.get("type") == "input_json_delta":
                        if current_tool_block:
                            current_tool_block["function"]["arguments"] += delta.get("partial_json", "")

                elif event_type == "content_block_stop":
                    if current_tool_block:
                        yield {"type": "tool_call", "tool_call": current_tool_block}
                        current_tool_block = None

            if current_tool_block:
                yield {"type": "tool_call", "tool_call": current_tool_block}

File: libs/sdk/src/agents/anthropic.rs
use crate::agents::traits::{AIProvider, StreamResponse};
use crate::agents::types::StreamChunk;
use crate::agents::utils::parse_anthropic_sse;
use crate::core::{Message, Role, ToolCall};
use async_stream::stream;
use async_trait::async_trait;
use futures::StreamExt;
use reqwest::Client;
use serde_json::{json, Value};
use std::collections::HashMap;
use std::sync::Arc;

pub struct AnthropicProvider {
    api_key: String,
    client: Client,
    base_url: String,
}

impl AnthropicProvider {
    pub fn new(api_key: String, base_url: Option<String>) -> Self {
        Self {
            api_key,
            client: Client::builder()
                .timeout(std::time::Duration::from_secs(60))
                .build()
                .unwrap_or_else(|_| Client::new()),
            base_url: base_url.unwrap_or_else(|| "https://api.anthropic.com".to_string()),
        }
    }
}

#[async_trait]
impl AIProvider for AnthropicProvider {
    fn name(&self) -> &str {
        "Anthropic"
    }

    async fn list_models(&self) -> Result<Vec<String>, anyhow::Error> {
        if let Some(models) = crate::utils::models::get_models_for_provider("anthropic") {
            return Ok(models);
        }
        // Anthropic doesn't have a public models endpoint in the same way OpenAI does that's easily accessible without specific permissions
        // Returning a common set of models
        Ok(vec![
            "claude-sonnet-4-5".to_string(),
            "claude-sonnet-4-20250514".to_string(),
            "claude-opus-4-20250514".to_string(),
            "claude-3-5-sonnet-20240620".to_string(),
            "claude-3-opus-20240229".to_string(),
            "claude-3-sonnet-20240229".to_string(),
            "claude-3-haiku-20240307".to_string(),
        ])
    }

    async fn ask(
        &self,
        messages: Arc<Vec<Message>>,
        model: &str,
        tools: Arc<Option<Vec<Value>>>,
        _thinking_level: Option<&str>,
    ) -> Result<StreamResponse, anyhow::Error> {
        let mut anthropic_messages = Vec::new();
        let mut system_prompt = String::new();

        for msg in messages.iter() {
            match msg.role {
                Role::System => {
                    if let Some(content) = &msg.content {
                        system_prompt.push_str(content);
                    }
                }
                Role::User => {
                    anthropic_messages.push(json!({
                        "role": "user",
                        "content": msg.content.as_deref().unwrap_or_default(),
                    }));
                }
                Role::Assistant => {
                    let mut content = Vec::new();
                    if let Some(t) = &msg.thought {
                        content.push(json!({ "type": "thought", "thought": t }));
                    }
                    if let Some(c) = &msg.content {
                        content.push(json!({ "type": "text", "text": c }));
                    }
                    if let Some(calls) = &msg.tool_calls {
                        for tc in calls {
                            let input: Value = serde_json::from_str(&tc.function.arguments).unwrap_or(json!({}));
                            content.push(json!({
                                "type": "tool_use",
                                "id": tc.id,
                                "name": tc.function.name,
                                "input": input,
                            }));
                        }
                    }
                    anthropic_messages.push(json!({
                        "role": "assistant",
                        "content": content,
                    }));
                }
                Role::Tool => {
                    anthropic_messages.push(json!({
                        "role": "user",
                        "content": [{
                            "type": "tool_result",
                            "tool_use_id": msg.tool_call_id.as_deref().unwrap_or_default(),
                            "content": msg.content.as_deref().unwrap_or_default(),
                        }],
                    }));
                }
            }
        }

        let mut body = json!({
            "model": model,
            "messages": anthropic_messages,
            "stream": true,
            "max_tokens": 16384,
        });

        if !system_prompt.is_empty() {
            body["system"] = json!(system_prompt);
        }

        if let Some(t) = tools.as_ref() {
            let mut anthropic_tools = Vec::new();
            for tool in t {
                if let Some(f) = tool.get("function") {
                    anthropic_tools.push(json!({
                        "name": f["name"],
                        "description": f["description"],
                        "input_schema": f["parameters"],
                    }));
                }
            }
            body["tools"] = json!(anthropic_tools);
        }

        let mut url = self.base_url.clone();
        if !url.ends_with("/v1/messages") {
            if url.ends_with("/v1/") || url.ends_with("/v1") {
                url = format!("{}/messages", url.trim_end_matches('/'));
            } else {
                url = format!("{}/v1/messages", url.trim_end_matches('/'));
            }
        }
        
        let response = self
            .client
            .post(&url)
            .header("x-api-key", &self.api_key)
            .header("anthropic-version", "2023-06-01")
            .header("content-type", "application/json")
            .json(&body)
            .send()
            .await?;

        let response = crate::utils::error::check_status(response).await?;

        let mut bytes_stream = response.bytes_stream();
        let mut buffer = String::new();
        let mut active_tool_calls: HashMap<usize, ToolCall> = HashMap::new();

        let s = stream! {
            while let Some(item) = bytes_stream.next().await {
                match item {
                    Ok(bytes) => {
                        let chunks = parse_anthropic_sse(&mut buffer, &mut active_tool_calls, &String::from_utf8_lossy(&bytes));
                        for chunk in chunks {
                            yield Ok(chunk);
                        }
                    }
                    Err(e) => yield Err(anyhow::Error::from(e)),
                }
            }
            yield Ok(StreamChunk::Done);
        };

        Ok(Box::pin(s))
    }
}
